Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Degree.lean

```diff
@@ -0,0 +1,253 @@
+      Nat.add_sub_cancel, Nat.mul_div_cancel_left _ two_pos]
+    ring1
+  · push_cast [expDegree, show (2 * n + 1) ^ 2 = 2 * (2 * n * (n + 1)) + 1 by ring1,
+      n.not_even_two_mul_add_one, Nat.add_sub_cancel, Nat.mul_div_cancel_left _ two_pos]
+    ring1
+
+private lemma expDegree_rec (m : ℕ) :
+    (expDegree (2 * (m + 3)) = 2 * expDegree (m + 2) + expDegree (m + 3) + expDegree (m + 5) ∧
+    expDegree (2 * (m + 3)) = expDegree (m + 1) + expDegree (m + 3) + 2 * expDegree (m + 4)) ∧
+    (expDegree (2 * (m + 2) + 1) =
+      expDegree (m + 4) + 3 * expDegree (m + 2) + (if Even m then 2 * 3 else 0) ∧
+    expDegree (2 * (m + 2) + 1) =
+      expDegree (m + 1) + 3 * expDegree (m + 3) + (if Even m then 0 else 2 * 3)) := by
+  push_cast [← @Nat.cast_inj ℤ, ← mul_left_cancel_iff_of_pos (b := (expDegree _ : ℤ)) two_pos,
+    mul_add, mul_left_comm (2 : ℤ)]
+  repeat rw [expDegree_cast <| by omega]
+  push_cast [Nat.even_add_one, ite_not, even_two_mul]
+  constructor <;> constructor <;> split_ifs <;> ring1
+
+private def expCoeff (n : ℕ) : ℤ :=
+  if Even n then n / 2 else n
+
+private lemma expCoeff_cast (n : ℕ) : (expCoeff n : ℚ) = if Even n then (n / 2 : ℚ) else n := by
+  rcases n.even_or_odd' with ⟨n, rfl | rfl⟩ <;> simp [expCoeff, n.not_even_two_mul_add_one]
+
+private lemma expCoeff_rec (m : ℕ) :
+    (expCoeff (2 * (m + 3)) =
+      expCoeff (m + 2) ^ 2 * expCoeff (m + 3) * expCoeff (m + 5) -
+        expCoeff (m + 1) * expCoeff (m + 3) * expCoeff (m + 4) ^ 2) ∧
+    (expCoeff (2 * (m + 2) + 1) =
+      expCoeff (m + 4) * expCoeff (m + 2) ^ 3 * (if Even m then 4 ^ 2 else 1) -
+        expCoeff (m + 1) * expCoeff (m + 3) ^ 3 * (if Even m then 1 else 4 ^ 2)) := by
+  push_cast [← @Int.cast_inj ℚ, expCoeff_cast, even_two_mul, m.not_even_two_mul_add_one,
+    Nat.even_add_one, ite_not]
+  constructor <;> split_ifs <;> ring1
+
+private lemma natDegree_coeff_preΨ' (n : ℕ) :
+    (W.preΨ' n).natDegree ≤ expDegree n ∧ (W.preΨ' n).coeff (expDegree n) = expCoeff n := by
+  induction n using normEDSRec with
+  | zero => simpa only [preΨ'_zero] using ⟨by rfl, Int.cast_zero.symm⟩
+  | one => simpa only [preΨ'_one] using ⟨natDegree_one.le, coeff_one_zero.trans Int.cast_one.symm⟩
+  | two => simpa only [preΨ'_two] using ⟨natDegree_one.le, coeff_one_zero.trans Int.cast_one.symm⟩
+  | three => simpa only [preΨ'_three] using ⟨W.natDegree_Ψ₃, W.coeff_Ψ₃ ▸ (Int.cast_ofNat 3).symm⟩
+  | four =>
+    simpa only [preΨ'_four] using ⟨W.natDegree_preΨ₄, W.coeff_preΨ₄ ▸ (Int.cast_ofNat 2).symm⟩
+  | even m h₁ h₂ h₃ h₄ h₅ =>
+    let dm {m n p q} : _ → _ → (p * q : R[X]).natDegree ≤ m + n := natDegree_mul_le_of_le
+    let dp {m n p} : _ → (p ^ n : R[X]).natDegree ≤ n * m := natDegree_pow_le_of_le n
+    let cm {m n p q} : _ → _ → (p * q : R[X]).coeff (m + n) = _ := coeff_mul_of_natDegree_le
+    let cp {m n p} : _ → (p ^ m : R[X]).coeff (m * n) = _ := coeff_pow_of_natDegree_le
```

Comment:
I'm wondering why these are necessary, which aren't there in my branch? I'll check out the branch and take a look myself later ...

Reply:
These shorten the code by a bit! Note that I killed off `natDegree/coeff_even/odd` by doing this.

Reply:
I see, you got rid of `natDegree/coeff_even/odd`! That's really remarkable 😀